### PR TITLE
Adds latin1 support to ANCM In-proc

### LINKED ISSUE #22675: Allow ANCM in-process to interpret request headers as Latin1 encoded
Similar issue with Kestrel: https://github.com/dotnet/aspnetcore/issues/17399

Related PR for the fix in Kestrel: https://github.com/dotnet/aspnetcore/pull/18255

This is a very similar ask that we had for Latin1 support in Kestrel. 

In a 3.1 patch we should provide:
* A config setting to change the default encoding to Latin1
* When enabled, ANCM will widen each incoming byte to a UTF-16 character and build a `string` from that set of characters, **instead** of interpreting bytes as UTF-8.
* We will continue to reject control characters from the ASCII set (0x00-0x1F and 0x7F), but **not** reject control characters from the widened Latin1 set (because they would collide with other interpretations of this data, like UTF-8).

This would allow a consumer to reinterpret the data in this string as a UTF-8 sequence if desired.

In 5.0 we'll look at broader work to improve this experience. The 3.1 goals are scoped down to specific requests we've received.

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/dotnet/aspnetcore/issues/22675.

Few notes:
- I could add support for latin1 for HttpSys in this PR as well.
- This PR is for master, for 3.1 I'll add an app context switch for latin1

### CODE REVIEW

File: src/Servers/IIS/IIS/src/IISServerOptions.cs [hunk 1]
```diff
@@ -65,5 +65,10 @@ public class IISServerOptions
                 _maxRequestBodySize = value;
             }
         }
+
+        /// <summary>
+        /// Treat request headers as Latin-1 or ISO/IEC 8859-1 instead of UTF-8.
+        /// </summary>
+        public bool Latin1RequestHeaders { get; set; } = false;
```
Comment (gfoidl) [inline src/Servers/IIS/IIS/src/IISServerOptions.cs:72]:
```suggestion
        public bool Latin1RequestHeaders { get; set; }
```
`false` is the default, or do you want to put explicit emphasis on this?

File: src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs [hunk 2]
```diff
@@ -3,24 +3,87 @@
...
 
 namespace Microsoft.AspNetCore.HttpSys.Internal
 {
     internal static class HeaderEncoding
     {
-        // It should just be ASCII or ANSI, but they break badly with un-expected values. We use UTF-8 because it's the same for
-        // ASCII, and because some old client would send UTF8 Host headers and expect UTF8 Location responses
-        // (e.g. IE and HttpWebRequest on intranets).
-        private static Encoding Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
+        private static Encoding DefaultEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
```
Comment (gfoidl) [inline src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:12]:
```suggestion
        private static readonly Encoding DefaultEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
```

File: src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs [hunk 3]
```diff
@@ -3,24 +3,87 @@
...
+            {
+                return GetLatin1StringNonNullCharacters(new Span<byte>(pBytes, byteCount));
+            }
+            else
+            {
+                return GetAsciiOrUTF8StringNonNullCharacters(new Span<byte>(pBytes, byteCount));
+            }
+        }
+
+        private static unsafe string GetLatin1StringNonNullCharacters(this Span<byte> span)
```
Comment (gfoidl) [inline src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:26]:
```suggestion
        private static unsafe string GetLatin1StringNonNullCharacters(Span<byte> span)
```
It's not used as extension method.

File: src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs [hunk 4]
```diff
@@ -3,24 +3,87 @@
...
+                {
+                    // null characters are considered invalid
+                    throw new InvalidOperationException();
+                }
+            }
+
+            return resultString;
+        }
+
+        private static unsafe string GetAsciiOrUTF8StringNonNullCharacters(this Span<byte> span)
```
Comment (gfoidl) [inline src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:49]:
```suggestion
        private static unsafe string GetAsciiOrUTF8StringNonNullCharacters(Span<byte> span)
```

File: src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs [hunk 5]
```diff
@@ -3,24 +3,87 @@
...
-        private static Encoding Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
+        private static Encoding DefaultEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
 
-        internal static unsafe string GetString(byte* pBytes, int byteCount)
+        internal static unsafe string GetString(byte* pBytes, int byteCount, bool useLatin1)
         {
-            return Encoding.GetString(new ReadOnlySpan<byte>(pBytes, byteCount));
+            if (useLatin1)
+            {
+                return GetLatin1StringNonNullCharacters(new Span<byte>(pBytes, byteCount));
```
Comment (gfoidl) [inline src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:18]: What's the point of creating the span, which is then "decomposed" to the pointer and length?
I would just pass the pointer/length down, as it's unsafe anyway.

File: src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs [hunk 6]
```diff
@@ -3,24 +3,87 @@
...
+        {
+            if (span.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            var resultString = new string('\0', span.Length);
+
+            fixed (char* output = resultString)
+            fixed (byte* buffer = span)
```
Comment (gfoidl) [inline src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:36]:
```suggestion
            fixed (byte* buffer = &MemoryMarshal.GetReference(span))
```
As we know that span isn't empty, we can safely go the "fast-path" (without another empty-check).

File: src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs [hunk 7]
```diff
@@ -3,24 +3,87 @@
...
+            if (span.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            var resultString = new string('\0', span.Length);
+            fixed (char* output = resultString)
+            fixed (byte* buffer = span)
+            {
+                // This version if AsciiUtilities returns null if there are any null (0 byte) characters
```
Comment (ilmax) [inline src/Shared/HttpSys/RequestProcessing/HeaderEncoding.cs:60]: nit: This version **of** AsciiUtilities

**Comment by Tratcher** [comment]
Should we close this master PR in favor of https://github.com/dotnet/aspnetcore/issues/17400?

**Comment by jkotalik** [comment]
yeah.

**Pull request closed without merging**